File: ais_decoder.py
import asyncio
import logging
from datetime import datetime, timezone

from pyais import decode

from config import OWN_MMSI

logger = logging.getLogger(__name__)

# AIS ship type ranges → human-readable categories
SHIP_TYPE_MAP = {
    range(20, 30): "Wing in Ground",
    range(30, 36): "Fishing/Towing/Dredging",
    range(36, 40): "Sailing/Pleasure",
    range(40, 50): "High Speed Craft",
    range(50, 60): "Special Craft",
    range(60, 70): "Passenger",
    range(70, 80): "Cargo",
    range(80, 90): "Tanker",
    range(90, 100): "Other",
}


def get_ship_category(ship_type: int | None) -> str:
    if ship_type is None:
        return "Unknown"
    for type_range, category in SHIP_TYPE_MAP.items():
        if ship_type in type_range:
            return category
    return "Other"


def decode_ais_message(sentence: str) -> dict | None:
    """Decode a single NMEA sentence. Returns a dict or None if not decodable."""
    try:
        decoded = decode(sentence)
        msg = decoded.asdict()

        result = {
            "mmsi": msg.get("mmsi"),
            "msg_type": msg.get("msg_type"),
            "is_own_vessel": msg.get("mmsi") == OWN_MMSI,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }

        # Position reports (msg types 1, 2, 3, 18, 19)
        lat = msg.get("lat")
        lon = msg.get("lon")
        if lat is not None and lon is not None:
            # AIS uses 91.0 / 181.0 as "not available"
            if abs(lat) <= 90.0 and abs(lon) <= 180.0:
                result["lat"] = round(lat, 6)
                result["lon"] = round(lon, 6)

        sog = msg.get("speed")
        if sog is not None and sog < 102.3:  # 102.3 = not available
            result["sog"] = round(sog, 1)

        cog = msg.get("course")
        if cog is not None and cog < 360.0:
            result["cog"] = round(cog, 1)

        heading = msg.get("heading")
        if heading is not None and heading < 360:
            result["heading"] = heading

        # Static data (msg types 5, 24)
        for field in ("shipname", "destination", "ship_type", "to_bow", "to_stern", "to_port", "to_starboard"):
            val = msg.get(field)
            if val is not None:
                result[field] = val

        # Compute length/beam from dimension fields
        if "to_bow" in result and "to_stern" in result:
            result["length"] = result["to_bow"] + result["to_stern"]
        if "to_port" in result and "to_starboard" in result:
            result["beam"] = result["to_port"] + result["to_starboard"]

        if result.get("ship_type") is not None:
            result["ship_category"] = get_ship_category(result["ship_type"])

        if result["mmsi"] is None:
            return None

        return result

    except Exception as e:
        logger.debug(f"Failed to decode: {sentence!r} — {e}")
        return None


async def ais_decoder(raw_queue: asyncio.Queue, decoded_queue: asyncio.Queue):
    """Read raw NMEA sentences, decode them, and push structured data to decoded_queue."""
    # Buffer for multi-part messages
    fragment_buffer: dict[tuple, list[str]] = {}

    while True:
        sentence = await raw_queue.get()

        try:
            # Check if multi-part message
            parts = sentence.split(",")
            if len(parts) >= 3 and parts[0] in ("!AIVDM", "!AIVDO"):
                total_frags = int(parts[1])
                frag_num = int(parts[2])

                if total_frags == 1:
                    # Single-part message — decode directly
                    result = decode_ais_message(sentence)
                    if result:
                        await decoded_queue.put(result)
                else:
                    # Multi-part: buffer fragments
                    seq_id = parts[3] if parts[3] else "0"
                    key = (parts[0], seq_id)

                    if frag_num == 1:
                        fragment_buffer[key] = [sentence]
                    elif key in fragment_buffer:
                        fragment_buffer[key].append(sentence)

                        if frag_num == total_frags:
                            # All fragments received — decode combined
                            combined = "\n".join(fragment_buffer.pop(key))
                            result = decode_ais_message(combined)
                            if result:
                                await decoded_queue.put(result)

        except Exception as e:
            logger.debug(f"Decoder error for {sentence!r}: {e}")

File: static/js/ais-decoder.js
/**
 * Browser-side AIS decoder for raw !AIVDM/!AIVDO sentences.
 * Output matches the shape aisstream.js parseAISStreamMessage() produces.
 */

const AISDecoder = (() => {

    function charToSixBit(c) {
        const v = c.charCodeAt(0);
        if (v >= 48 && v <= 87) return v - 48;
        if (v >= 96 && v <= 119) return v - 56;
        return 0;
    }

    function decodeBits(payload) {
        const bits = [];
        for (let i = 0; i < payload.length; i++) {
            const v = charToSixBit(payload[i]);
            for (let b = 5; b >= 0; b--) bits.push((v >> b) & 1);
        }
        return bits;
    }

    function getUint(bits, start, len) {
        let v = 0;
        for (let i = start; i < start + len && i < bits.length; i++) {
            v = (v << 1) | bits[i];
        }
        return v;
    }

    function getInt(bits, start, len) {
        let v = getUint(bits, start, len);
        if (bits[start] === 1) v -= (1 << len);
        return v;
    }

    function getText(bits, start, len) {
        const chars = [];
        for (let i = start; i < start + len; i += 6) {
            let c = getUint(bits, i, 6);
            if (c < 32) c += 64;
            chars.push(String.fromCharCode(c));
        }
        return chars.join('').replace(/@+$/, '').trim();
    }

    function decodeType123(bits) {
        const mmsi = getUint(bits, 8, 30);
        const status = getUint(bits, 38, 4);
        const rot = getInt(bits, 42, 8);
        const sog = getUint(bits, 50, 10) / 10;
        const lon = getInt(bits, 61, 28) / 600000;
        const lat = getInt(bits, 89, 27) / 600000;
        const cog = getUint(bits, 116, 12) / 10;
        const heading = getUint(bits, 128, 9);

        if (Math.abs(lat) > 90 || Math.abs(lon) > 180) return null;
        if (lat === 0 && lon === 0) return null;

        return {
            mmsi, msg_type: getUint(bits, 0, 6),
            lat: Math.round(lat * 1e6) / 1e6,
            lon: Math.round(lon * 1e6) / 1e6,
            sog: sog < 102.3 ? Math.round(sog * 10) / 10 : null,
            cog: cog < 360 ? Math.round(cog * 10) / 10 : null,
            heading: heading < 360 ? heading : null,
            nav_status: status,
        };
    }

    function decodeType5(bits) {
        const mmsi = getUint(bits, 8, 30);
        const name = getText(bits, 112, 120);
        const shipType = getUint(bits, 232, 8);
        const dimA = getUint(bits, 240, 9);
        const dimB = getUint(bits, 249, 9);
        const dimC = getUint(bits, 258, 6);
        const dimD = getUint(bits, 264, 6);
        const dest = getText(bits, 302, 120);

        const result = { mmsi, msg_type: 5, name: name || null };
        if (shipType) {
            result.ship_type = shipType;
            result.ship_category = getShipCategoryLocal(shipType);
        }
        if (dimA + dimB > 0) result.length = dimA + dimB;
        if (dimC + dimD > 0) result.beam = dimC + dimD;
        if (dest) result.destination = dest;
        return result;
    }

    function decodeType18(bits) {
        const mmsi = getUint(bits, 8, 30);
        const sog = getUint(bits, 46, 10) / 10;
        const lon = getInt(bits, 57, 28) / 600000;
        const lat = getInt(bits, 85, 27) / 600000;
        const cog = getUint(bits, 112, 12) / 10;
        const heading = getUint(bits, 124, 9);

        if (Math.abs(lat) > 90 || Math.abs(lon) > 180) return null;
        if (lat === 0 && lon === 0) return null;

        return {
            mmsi, msg_type: 18,
            lat: Math.round(lat * 1e6) / 1e6,
            lon: Math.round(lon * 1e6) / 1e6,
            sog: sog < 102.3 ? Math.round(sog * 10) / 10 : null,
            cog: cog < 360 ? Math.round(cog * 10) / 10 : null,
            heading: heading < 360 ? heading : null,
        };
    }

    function decodeType19(bits) {
        const result = decodeType18(bits);
        if (!result) return null;
        result.msg_type = 19;
        result.name = getText(bits, 143, 120) || null;
        const shipType = getUint(bits, 263, 8);
        if (shipType) {
            result.ship_type = shipType;
            result.ship_category = getShipCategoryLocal(shipType);
        }
        return result;
    }

    function decodeType24(bits) {
        const mmsi = getUint(bits, 8, 30);
        const partNo = getUint(bits, 38, 2);
        if (partNo === 0) {
            return { mmsi, msg_type: 24, name: getText(bits, 40, 120) || null };
        } else if (partNo === 1) {
            const shipType = getUint(bits, 40, 8);
            const result = { mmsi, msg_type: 24 };
            if (shipType) {
                result.ship_type = shipType;
                result.ship_category = getShipCategoryLocal(shipType);
            }
            return result;
        }
        return null;
    }

    const SHIP_TYPE_MAP = [
        [20, 30, 'Wing in Ground'], [30, 36, 'Fishing/Towing/Dredging'],
        [36, 40, 'Sailing/Pleasure'], [40, 50, 'High Speed Craft'],
        [50, 60, 'Special Craft'], [60, 70, 'Passenger'],
        [70, 80, 'Cargo'], [80, 90, 'Tanker'], [90, 100, 'Other'],
    ];

    function getShipCategoryLocal(t) {
        for (const [lo, hi, cat] of SHIP_TYPE_MAP) {
            if (t >= lo && t < hi) return cat;
        }
        return 'Other';
    }

    const DECODERS = {
        1: decodeType123, 2: decodeType123, 3: decodeType123,
        5: decodeType5, 18: decodeType18, 19: decodeType19, 24: decodeType24,
    };

    const fragmentBuffer = new Map();
    const FRAGMENT_TIMEOUT = 10000;

    function processSentence(sentence) {
        const starIdx = sentence.indexOf('*');
        const body = starIdx >= 0 ? sentence.slice(1, starIdx) : sentence.slice(1);
        const fields = body.split(',');

        // fields[0] = talker (AIVDM/AIVDO), rest follow standard order
        const fragCount = parseInt(fields[1]);
        const fragNum = parseInt(fields[2]);
        const seqId = fields[3] || '';
        const channel = fields[4] || '';
        const payload = fields[5] || '';
        const fillBits = parseInt(fields[6]) || 0;

        if (fragCount === 1) {
            return decode(payload, fillBits);
        }

        const key = `${channel}:${seqId}`;
        if (fragNum === 1) {
            fragmentBuffer.set(key, { parts: [payload], count: fragCount, time: Date.now() });
        } else {
            const buf = fragmentBuffer.get(key);
            if (!buf) return null;
            buf.parts.push(payload);
            if (buf.parts.length === buf.count) {
                fragmentBuffer.delete(key);
                return decode(buf.parts.join(''), fillBits);
            }
        }

        for (const [k, v] of fragmentBuffer) {
            if (Date.now() - v.time > FRAGMENT_TIMEOUT) fragmentBuffer.delete(k);
        }

        return null;
    }

    function decode(payload, fillBits) {
        const bits = decodeBits(payload);
        const msgType = getUint(bits, 0, 6);
        const decoder = DECODERS[msgType];
        if (!decoder) return null;
        const result = decoder(bits);
        if (result) {
            result.timestamp = new Date().toISOString();
            result.is_own_vessel = false;
        }
        return result;
    }

    return { processSentence, decode };
})();
